src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (let j = 0; j < action.keysToRemove.length; j++) {
+          const key = action.keysToRemove[j];
+
+          delete currentState[key];
+        }
+        break;
+
+      case 'clear':
+        currentState = {};
+        break;
+    }
```

Comment:
Consider adding a default case in your switch statement to handle any unexpected action types. This will improve error handling and ensure that your function behaves predictably even with incorrect input.